x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java

```diff
@@ -352,10 +395,118 @@ public Relation isFieldWithinQuery(
-            return delegateFieldType().fielddataBuilder(fullyQualifiedIndexName);
+        public ValuesSourceType getValuesSourceType() {
+            return AggregateMetricsValuesSourceType.AGGREGATE_METRIC;
         }
 
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            return new IndexFieldData.Builder() {
+                @Override
+                public IndexFieldData<?> build(
+                    IndexSettings indexSettings,
+                    MappedFieldType fieldType,
+                    IndexFieldDataCache cache,
+                    CircuitBreakerService breakerService,
+                    MapperService mapperService
+                ) {
+                    return new IndexAggregateDoubleMetricFieldData(indexSettings.getIndex(), fieldType.name()) {
+                        @Override
+                        public LeafAggregateDoubleMetricFieldData load(LeafReaderContext context) {
+                            return new LeafAggregateDoubleMetricFieldData() {
+                                @Override
+                                public SortedNumericDoubleValues getAggregateMetricValues(final Metric metric) throws IOException {
+                                    try {
+                                        final SortedNumericDocValues values = DocValues.getSortedNumeric(
+                                            context.reader(),
+                                            subfieldName(fieldName, metric)
+                                        );
+
+                                        return new SortedNumericDoubleValues() {
+                                            @Override
+                                            public int docValueCount() {
+                                                return values.docValueCount();
+                                            }
+
+                                            @Override
+                                            public boolean advanceExact(int doc) throws IOException {
+                                                return values.advanceExact(doc);
+                                            }
+
+                                            @Override
+                                            public double nextValue() throws IOException {
+                                                long v = values.nextValue();
+                                                if (metric == Metric.value_count) {
+                                                    // Only value_count metrics are encoded as integers
+                                                    return v;
+                                                } else {
+                                                    // All other metrics are encoded as doubles
+                                                    return NumericUtils.sortableLongToDouble(v);
+                                                }
```

Comment:
When parsing and inserting a `value_count` metric we took the decision to treat as `integer` (see https://github.com/elastic/elasticsearch/blob/32be5aae223a5e876cc4aa1d6ca7890b01cb194c/x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java#L200)

So when we aggregate documents, we must treat `value_count` sub-fields as longs and not as doubles. This part of the code is a bit ugly. Is there any better way to do this?